### PR TITLE
[PT-162549243] Add extension commands for win32

### PR DESCRIPTION
**Pull request by tolbrino**
NOTE: Based on #2139

Two bash extension command scripts have been re-written to move most of the logic into escripts, which are more portable.

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -16,6 +17,11 @@
 %% built (!) though the explicit copying in the `relx` overlay fails.
 %% Addressing such a failure consists of updating the content of
 %% `otp_patches/`.
+%%
+%% The project uses a modified rebar3 which is built using this commit:
+%%
+%% https://github.com/tolbrino/rebar3/commit/7b6a07597eb347294c0588d8435aa647a19a6bce
```
Comment (lucafavatella) [inline rebar.config]: This is prone to getting out-of-date. I would favour detailed commit message adding rebar3 e.g. https://github.com/aeternity/aeternity/commit/3493e9ec94faa34bf4d57493c93e168bcf00e765

File: Makefile [hunk 2]
```diff
@@ -325,7 +325,7 @@ multi-build: dev1-build
 .SECONDEXPANSION:
 
 internal-compile-deps: $$(KIND)
-	@$(REBAR) as $(KIND) compile --deps-only
+	@$(REBAR) as $(KIND) compile -d
```
Comment (lucafavatella) [inline Makefile]: I see [documented](https://github.com/erlang/rebar3/blob/3.9.0/src/rebar_prv_compile.erl#L33-L34). Ok.

File: rebar.config.script [hunk 3]
```diff
@@ -20,13 +20,34 @@ FilterDepsRocksDb = fun(Deps) ->
...
                     end,
 
+OverlayReplacements = [
+                       "bin/epoch",
+                       "bin/extensions/check_config",
+                       "bin/extensions/keys_gen",
+                       "bin/extensions/peer_key",
+                       "bin/extensions/messages_hash",
+                       "bin/extensions/export_chain"
+                      ],
```
Comment (lucafavatella) [inline rebar.config.script]: This list seems prone to get out of sync. Would it be possible, at build time, to cause error if any of the following constraints is not adhered to:
* For each item X of the `OverlayReplacements`, there must be a `{copy, Src0, Dest0}` with `Dest0` equal to X.
* Each item among the `{copy, Src0, Dest0}` items where `Dest0` starts with `"bin/extensions/` must be such that `Dest0` is among `OverlayReplacements`.

File: ci/appveyor/package.cfg [hunk 4]
```diff
@@ -4,16 +4,21 @@ packages = {pkg_prefix}libsodium
 
 filename_stub = aeternity-windows
 
-display_name = aeternity for Windows
+display_name = Aeternity for Windows
 
-description = The full aeternity node for Windows.
+description = The full Aeternity node for Windows.
 
 version = 1.5.0-alpha.1
```
Comment (lucafavatella) [inline ci/appveyor/package.cfg]: I understand the plan is to consider this for the minerva branch. Could you please rebase on minerva branch?

File: ci/appveyor/package.cfg [hunk 5]
```diff
@@ -4,16 +4,21 @@ packages = {pkg_prefix}libsodium
...
 
 url = https://github.com/aeternity/aeternity
 
 launchers =
     aeternity.desktop
+    keys_gen.util
+    check_config.util
+    export_chain.util
+    messages_hash.util
+    peer_key.util
```
Comment (lucafavatella) [inline ci/appveyor/package.cfg]: Other list prone to get out of sync ... is there any automated check? Or a smoke test that exercise at least once each of the lists?

Comment (lucafavatella) [reply ci/appveyor/package.cfg]: I see test.bat commented in CI.

Comment (lucafavatella) [reply ci/appveyor/package.cfg]: Wrt extensions themselves, I see two ways that those could be tested both on Win and Linux/macOS:
* Integration tests (`make ct`)
* Package smoke test ([`make python-release-test`](https://github.com/aeternity/aeternity/blob/e58cfd5c5677f67828d67c1e8364b7b0b1b9446b/.circleci/config.yml#L130-L140))

**Comment by lucafavatella** [comment]
This seems quite a large PR, may be prone to delayed merge. I wonder if it could be rebased on minerva branch then split in:
* Elements that can go in without changing functional behaviour.
* Elements that are enable win (rebar3, win-specific changes).

Though could be kept as one.

And I am concerned of test (at least smoke) coverage on Win.

**Comment by tolbrino** [comment]
@lucafavatella I rebased on `minerva` and pulled out the rebar update into a separate PR #2139 .

Tests are following.

**Comment by lucafavatella** [comment]
I see this is based (and includes) https://github.com/aeternity/aeternity/pull/2139

File: ci/appveyor/package.cfg (hunk 5 above)
Comment (tolbrino) [reply ci/appveyor/package.cfg]: I added a python-based test.

**Comment by tolbrino** [comment]
@lucafavatella @velzevur I've updated the PR based on your feedback. Could you re-review it?

File: apps/aeutils/priv/extensions/export_chain.sh [hunk 6]
```diff
@@ -1,59 +1,11 @@
-#!/bin/bash
```
Comment (lucafavatella) [inline apps/aeutils/priv/extensions/export_chain.sh]: I did not find automated tests for this script. Can you please confirm that it still works.

File: apps/aeutils/priv/extensions/get_peer_key.sh [hunk 7]
```diff
@@ -1,39 +1,11 @@
-#!/bin/bash
```
Comment (lucafavatella) [inline apps/aeutils/priv/extensions/get_peer_key.sh]: I see there is a test for this script. Ok.

File: apps/aeutils/priv/extensions/export_chain.sh [hunk 8]
```diff
@@ -1,59 +1,11 @@
...
-        ok)
-            echo "$RESULT_PAYLOAD blocks exported to $FILE_PATH"
-            exit 0
-            ;;
-    esac
-fi
-
-echo "ERROR: $RESULT"
-exit 1
+exit $?
```
Comment (lucafavatella) [inline apps/aeutils/priv/extensions/export_chain.sh]: I wonder - why not running `exec relx_escript ...` directly?

File: apps/aeutils/priv/extensions/get_peer_key.sh [hunk 9]
```diff
@@ -1,39 +1,11 @@
...
+APPS_VSN=${REL_VSN:?}
 
-RC=$?
-if [ $RC -eq 0 ]; then
-    exit 1
-fi
+relx_escript "lib/aeutils-${APPS_VSN:?}/priv/get_peer_key" "$NAME_TYPE" "$NAME" -setcookie "$COOKIE" $*
 
-exit 0
+exit $?
```
Comment (lucafavatella) [inline apps/aeutils/priv/extensions/get_peer_key.sh]: I wonder - why not running `exec relx_escript ...` directly?

File: apps/aeutils/priv/extensions/messages_hash.sh [hunk 10]
```diff
@@ -1,10 +1,11 @@
...
 
 # Wrapper to run messages_hash
 
 PATH=$BINDIR:$PATH
 export ERL_LIBS=$ROOTDIR/lib
 APPS_VSN=${REL_VSN:?}
+
 "$ERTS_DIR/bin/escript" "$ROOTDIR/lib/aeutils-${APPS_VSN:?}/priv/messages_hash" "$*"
 
 exit $?
```
Comment (lucafavatella) [inline apps/aeutils/priv/extensions/messages_hash.sh]: I wonder - why not running `exec ...` directly?

File: system_test/aest_commands_SUITE.erl [hunk 11]
```diff
@@ -66,7 +66,7 @@ commands_(OpsBin, Cfg) ->
         "Installed versions:[\r\n]*\\*[ \t]*[0-9\\.\\-a-z]*[ \t]*permanent[\r\n]*")),
 
     {0, Output2} = run_cmd_in_node_dir(node1, [OpsBin, "peer_key"], Cfg),
-    {ok, PeerKey} = aehttp_api_encoder:safe_decode(peer_pubkey, list_to_binary(Output2)),
+    {ok, PeerKey} = aehttp_api_encoder:safe_decode(peer_pubkey, list_to_binary(string:trim(Output2, trailing, "\n"))),
```
Comment (lucafavatella) [inline system_test/aest_commands_SUITE.erl]: Does this mean that the script now returns one newline more? Please confirm with @dincho that it is ok. (I may be easy to keep previous behaviour.)

File: py/tests/package_win32.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+                        help='Working directory for testing. It must exist and be empty.')
+    parser.add_argument('--packagespecfile', type=argparse.FileType('r'), required=True,
+                        help='package specification file. It must exist.')
+
+    args = parser.parse_args()
+    return (args.workdir, args.packagespecfile)
+
+def finish_test(test_failed):
+    if test_failed:
+        sys.exit('FAILED')
```
Comment (lucafavatella) [inline py/tests/package_win32.py]: I see non-zero exit code. Ok.

File: rebar.config (hunk 1 above)
Comment (tolbrino) [reply rebar.config]: removed

File: rebar.config.script (hunk 3 above)
Comment (tolbrino) [reply rebar.config.script]: made generic

File: ci/appveyor/package.cfg (hunk 4 above)
Comment (tolbrino) [reply ci/appveyor/package.cfg]: done

File: apps/aeutils/priv/extensions/export_chain.sh (hunk 6 above)
Comment (tolbrino) [reply apps/aeutils/priv/extensions/export_chain.sh]: It works. Tested on Windows and Unix.

File: apps/aeutils/priv/extensions/export_chain.sh (hunk 8 above)
Comment (tolbrino) [reply apps/aeutils/priv/extensions/export_chain.sh]: I tried to stay close to the previous style of bash scripts.

File: apps/aeutils/priv/extensions/export_chain.sh (hunk 6 above)
Comment (lucafavatella) [reply apps/aeutils/priv/extensions/export_chain.sh]: Thanks.

**Pull request merged**